Libs/Mesh/MeshUtils.cpp

```diff
@@ -43,6 +57,125 @@ const vtkSmartPointer<vtkMatrix4x4> MeshUtils::createICPTransform(const vtkSmart
+	vtkSmartPointer<vtkDataArray> dataArray = points->GetData();
+	
+  int numVertices = points->GetNumberOfPoints();
+	
+	Eigen::MatrixXd Vref_new(numVertices, 3);
+	
+	for(int i=0; i<numVertices;i++){
+		Vref_new(i, 0) = dataArray->GetComponent(i, 0);
+		Vref_new(i, 1) = dataArray->GetComponent(i, 1);
+		Vref_new(i, 2) = dataArray->GetComponent(i, 2);
+	}
+  return Vref_new;
+}
+
+Eigen::MatrixXi MeshUtils::distilFaceInfo(Mesh mesh){
+  vtkSmartPointer<vtkPolyData> poly_data = mesh.getVTKMesh();
+  int numFaces = poly_data->GetNumberOfCells();
+  Eigen::MatrixXi Fref_new(numFaces, 3);
+	
+	vtkSmartPointer<vtkIdList> cellIdList =
+      vtkSmartPointer<vtkIdList>::New();
+	
+	for(int j = 0; j < numFaces; j++){
+		poly_data->GetCellPoints(j, cellIdList);
+		Fref_new(j, 0) = cellIdList->GetId(0);
+		Fref_new(j, 1) = cellIdList->GetId(1);
+		Fref_new(j, 2) = cellIdList->GetId(2);
+	}
+  return Fref_new;
+}
+
+
+Eigen::MatrixXd MeshUtils::generateWarpMatrix(Eigen::MatrixXd TV , Eigen::MatrixXi TF, Eigen::MatrixXd Vref){
+  
+  Eigen::MatrixXd W;
+  Eigen::VectorXi b;
+	{
+		Eigen::VectorXi J = Eigen::VectorXi::LinSpaced(TV.rows(),0,TV.rows()-1);
+		Eigen::VectorXd sqrD;
+		Eigen::MatrixXd _2;
+    //std::cout<<"Finding closest points..."<<std::endl;
+		igl::point_mesh_squared_distance(Vref,TV,J,sqrD,b,_2);
+    //std::cout << "sqrd " <<sqrD.minCoeff() <<std::endl;
+	}
+  // force perfect positioning, rather have popping in low-res than high-res.
+  // The correct/elaborate thing to do is express original low.V in terms of
+  // linear interpolation (or extrapolation) via elements in (high.V,high.F)
+  igl::slice(TV,b,1,Vref);
+  // list of points --> list of singleton lists
+  std::vector<std::vector<int> > S;
+  igl::matrix_to_list(b,S);
+  //std::cout<<"Computing weights for "<<b.size()<<
+    //" handles at "<<TV.rows()<<" vertices..."<<std::endl;
+  // Technically k should equal 3 for smooth interpolation in 3d, but 2 is
+  // faster and looks OK
+  const int k = 2;
+  igl::biharmonic_coordinates(TV,TF,S,k,W);
+  //std::cout<<"Reindexing..."<< std::endl;
+  //std::cout << W.rows() << " " << W.cols() << std::endl;
+  // Throw away interior tet-vertices, keep weights and indices of boundary
+  Eigen::VectorXi I,J;
+  igl::remove_unreferenced(TV.rows(),TF,I,J);
+  std::for_each(TF.data(),TF.data()+TF.size(),[&I](int & a){a=I(a);});
+  std::for_each(b.data(),b.data()+b.size(),[&I](int & a){a=I(a);});
+  igl::slice(Eigen::MatrixXd(TV),J,1,TV);
+  igl::slice(Eigen::MatrixXd(W),J,1,W);
+  //std::cout << "It's done!!" << std::endl;
+  return W;
+}
+
+Mesh MeshUtils::warpMesh(Eigen::MatrixXd movPts, Eigen::MatrixXd W, Eigen::MatrixXi Fref){
+  
+  int numVertices = W.rows();
+  int numFaces = Fref.rows();
+  // std::cout<< (Vcontrol_moving.rowwise() + Eigen::RowVector3d(0,0,0)) <<std::endl;
+	Eigen::MatrixXd Voutput = W * (movPts.rowwise() + Eigen::RowVector3d(0,0,0));
+	vtkSmartPointer<vtkPolyData> outmesh = vtkSmartPointer<vtkPolyData>::New();
```

Comment:
vtkSmartPointer<vtkPolyData> is defined as MeshType in Mesh.h

Reply:
Yes, I see that.  Personally, when I am writing VTK code, I would rather do the following:

```
vtkSmartPointer<vtkPolyData> outmesh
vtkSmartPointer<vtkPoints> outpoints
vtkSmartPointer<vtkCellArray> polys
..
```

Vs.:

```
MeshType outmesh
vtkSmartPointer<vtkPoints> outpoints
vtkSmartPointer<vtkCellArray> polys
...
```

For myself, when I see the latter, I think, MeshType?  What's that?  And then I have to look it up.

Reply:
I have the same feeling about this.

More broadly, for ShapeWorks we should pass around Mesh and Image rather than their internal data structures outside of the functions that need them. (We're already mostly doing this, but it's worth reiterating)

Reply:
Makes sense.